core/headerchain.go

```diff
@@ -566,15 +566,33 @@ type (
 // SetHead rewinds the local chain to a new head. Everything above the new head
 // will be deleted and the new one set.
 func (hc *HeaderChain) SetHead(head uint64, updateFn UpdateHeadBlocksCallback, delFn DeleteBlockContentCallback) {
+	hc.setHead(head, 0, updateFn, delFn)
+}
+
+// SetHeadWithTimestamp rewinds the local chain to a new head timestamp. Everything
+// above the new head will be deleted and the new one set.
+func (hc *HeaderChain) SetHeadWithTimestamp(time uint64, updateFn UpdateHeadBlocksCallback, delFn DeleteBlockContentCallback) {
+	hc.setHead(0, time, updateFn, delFn)
+}
+
+// setHead rewinds the local chain to a new head block or a head timestamp.
+// Everything above the new head will be deleted and the new one set.
+func (hc *HeaderChain) setHead(headBlock uint64, headTime uint64, updateFn UpdateHeadBlocksCallback, delFn DeleteBlockContentCallback) {
 	var (
 		parentHash common.Hash
 		batch      = hc.chainDb.NewBatch()
 		origin     = true
 	)
-	for hdr := hc.CurrentHeader(); hdr != nil && hdr.Number.Uint64() > head; hdr = hc.CurrentHeader() {
+	done := func(header *types.Header) bool {
+		if headBlock != 0 || headTime == 0 {
```

Comment:
Maybe we can simplify it by 

```
if headTime == 0 {
    return header.Number.Uint64() <= headBlock
}
return header.Time <= headTime
```

If headTime is zero, it must be block mode;
If headTime is non-zero, it must be time mode;

Reply:
> If headTime is zero, it must be block mode;
> If headTime is non-zero, it must be time mode;

If both are zero, it 's moot: we're going to the genesis. However: the genesis _number_ is `0`, but is it guaranteed that the genesis _time_ is zero? 

Regardless, I think the code _as is_ works (goes into numbermode at `0,0`), and so does @rjl493456442 's suggestion.

Reply:
If both are zero, it means we want to rewind to genesis, (still time is zero, block mode).

The only exception is we want to rewind by timestamp, but the target is 0(zero timestamp). 
I feel like it's meaningless to have this case handling.

Reply:
I changed it to

```go
	done := func(header *types.Header) bool {
		if headTime > 0 {
			return header.Time <= headTime
		}
		return header.Number.Uint64() <= headBlock
	}
```

Seems cleaner to reason about but it's the exact same code as Gary's suggestion

Reply:
It's also analogous to the rest of the comparisons across the codebase.